### PR TITLE
Programatically filter out tool-induced fences from the sampler tool utility

### LINKED ISSUE #179: Invoke Kokkos::Fence programmatically from within a Kokkos tool  (focused on sampler)
Currently, a Kokkos Tool relies on an automatic global fencing to ensure that the Tool captures the appropriate state of its recording of state/diagnostic of kernel (e.g., a tool records time that the 20th invocation of a parallel_for() in a Kokkos program took). 

For some Kokkos Tools, like the Kokkos Sampler utility Tool where the kernel callback is invoked only periodically or with a probability: the tool should induce a fence _only_ for the kernel invocation that actually invokes recording of profiling/debugging data gathered by the tool. Doing so is important especially as a Tool like the sampler will often in practice be using very low sampling rates like 1%. If not done for sampling when sampling rate is at the example 1%, this will create large unnecessary tool-induced overheads coming from the other 99% of non-sampled kernels. 

The proposed solution is to use the tool profiling hook from kokkos-core (the same interface used for global auto-fencing) to 'conditionally force' (prescribe during Kokkos program's execution given a condition) a Kokkos::fence from Kokkos core to be called only for the tool callback invocation for which profiling/debugging data of a kernel is actually gathered or logged. 

This Issue is currently targeted at the sampler, and it should be 
Restated as guideline for Kokkos Tools developers and Kokkos programmers within documentation for building one's own Kokkos Tools connector.

### PR DESCRIPTION
**Pull request by vlkale**
Fix #179.

### CODE REVIEW

**Comment by vlkale** [comment]
This is a example run and output on MacBook using g++ (default) and Kokkos serial host backend for stream, without and with Kokkos Tools global fencing. 
 
### Kokkos Tools Global Fence turned off (no auto-fencing)

```
s1017105ca:stream vlkale$  export KOKKOS_TOOLS_GLOBALFENCES=0;  ./stream-1000iters.exe 
-------------------------------------------------------------
Kokkos STREAM Benchmark
-------------------------------------------------------------
KokkosP: Next library to call: /Users/vlkale/Desktop/vlap/wk/code/softwareTech/kokkos-tools/vbld2/profiling/simple-kernel-timer/libkp_kernel_timer.dylib
KokkosP: Loading child library ..
KokkosP: Simple Kernel Timer Library Initialized (sequence is 1, version: 20211015)
KokkosP: Function Status:
KokkosP: begin-parallel-for:      yes
KokkosP: begin-parallel-scan:     yes
KokkosP: begin-parallel-reduce:   yes
KokkosP: end-parallel-for:        yes
KokkosP: end-parallel-scan:       yes
KokkosP: end-parallel-reduce:     yes
KokkosP: Sampling rate set to: 105
Reports fastest timing per kernel
Creating Views...
Memory Sizes:
- Array Size:    100000000
- Per Array:           800.00 MB
- Total:              2400.00 MB
Benchmark kernels will be performed for 1000 iterations.
-------------------------------------------------------------
Initializing Views...
Starting benchmarking...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 1 calling child-end function...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 2 calling child-end function...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 3 calling child-end function...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 4 calling child-end function...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 5 calling child-end function...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 6 calling child-end function...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 7 calling child-end function...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 8 calling child-end function...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 9 calling child-end function...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 10 calling child-end function...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 11 calling child-end function...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 12 calling child-end function...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 13 calling child-end function...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 14 calling child-end function...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 15 calling child-end function...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 16 calling child-end function...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 17 calling child-end function...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 18 calling child-end function...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 19 calling child-end function...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 20 calling child-end function...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 21 calling child-end function...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 22 calling child-end function...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 23 calling child-end function...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 24 calling child-end function...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 25 calling child-end function...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 26 calling child-end function...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 27 calling child-end function...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 28 calling child-end function...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 29 calling child-end function...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 30 calling child-end function...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 31 calling child-end function...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 32 calling child-end function...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 33 calling child-end function...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 34 calling child-end function...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 35 calling child-end function...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 36 calling child-end function...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 37 calling child-end function...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 38 calling child-end function...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 39 calling child-end function...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 40 calling child-end function...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 41 calling child-end function...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 42 calling child-end function...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 43 calling child-end function...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 44 calling child-end function...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 45 calling child-end function...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 46 calling child-end function...
Performing validation...
All solutions checked and verified.
-------------------------------------------------------------
Set                12559.15 MB/s
Copy               15953.07 MB/s
Scale              15780.36 MB/s
Add                17448.99 MB/s
Triad              17189.12 MB/s
-------------------------------------------------------------
KokkosP: Kernel timing written to /Users/vlkale/Desktop/vlap/wk/code/softwareTech/kokkos/benchmarks/stream/s1017105ca-28032.dat 
```


### Kokkos Tools Global Fence (auto-fencing) on


```
s1017105ca:stream vlkale$ export KOKKOS_TOOLS_GLOBALFENCES=1; ./stream-1000iters.exe 
-------------------------------------------------------------
Kokkos STREAM Benchmark
-------------------------------------------------------------
KokkosP: Next library to call: /Users/vlkale/Desktop/vlap/wk/code/softwareTech/kokkos-tools/vbld2/profiling/simple-kernel-timer/libkp_kernel_timer.dylib
KokkosP: Loading child library ..
KokkosP: Simple Kernel Timer Library Initialized (sequence is 1, version: 20211015)
KokkosP: Function Status:
KokkosP: begin-parallel-for:      yes
KokkosP: begin-parallel-scan:     yes
KokkosP: begin-parallel-reduce:   yes
KokkosP: end-parallel-for:        yes
KokkosP: end-parallel-scan:       yes
KokkosP: end-parallel-reduce:     yes
KokkosP: Sampling rate set to: 105
Reports fastest timing per kernel
Creating Views...
Memory Sizes:
- Array Size:    100000000
- Per Array:           800.00 MB
- Total:              2400.00 MB
Benchmark kernels will be performed for 1000 iterations.
-------------------------------------------------------------
Initializing Views...
Starting benchmarking...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 1 calling child-end function...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 2 calling child-end function...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 3 calling child-end function...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 4 calling child-end function...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 5 calling child-end function...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 6 calling child-end function...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 7 calling child-end function...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 8 calling child-end function...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 9 calling child-end function...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 10 calling child-end function...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 11 calling child-end function...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 12 calling child-end function...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 13 calling child-end function...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 14 calling child-end function...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 15 calling child-end function...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 16 calling child-end function...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 17 calling child-end function...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 18 calling child-end function...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 19 calling child-end function...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 20 calling child-end function...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 21 calling child-end function...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 22 calling child-end function...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 23 calling child-end function...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 24 calling child-end function...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 25 calling child-end function...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 26 calling child-end function...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 27 calling child-end function...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 28 calling child-end function...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 29 calling child-end function...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 30 calling child-end function...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 31 calling child-end function...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 32 calling child-end function...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 33 calling child-end function...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 34 calling child-end function...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 35 calling child-end function...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 36 calling child-end function...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 37 calling child-end function...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 38 calling child-end function...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 39 calling child-end function...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 40 calling child-end function...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 41 calling child-end function...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 42 calling child-end function...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 43 calling child-end function...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 44 calling child-end function...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 45 calling child-end function...
KokkosP: sample 1 calling child-begin function...
KokkosP: sample 46 calling child-end function...
Performing validation...
All solutions checked and verified.
-------------------------------------------------------------
Set                12224.61 MB/s
Copy               15412.82 MB/s
Scale              15549.36 MB/s
Add                16772.14 MB/s
Triad              16554.17 MB/s
-------------------------------------------------------------
KokkosP: Kernel timing written to /Users/vlkale/Desktop/vlap/wk/code/softwareTech/kokkos/benchmarks/stream/s1017105ca-28446.dat 
```

File: common/kokkos-sampler/kp_sampler_skip.cpp [hunk 1]
```diff
@@ -3,15 +3,19 @@
 #include <cstdlib>
 #include <cstring>
 #include <dlfcn.h>
+// #include <impl/Kokkos_Profiling_Interface.hpp>
 #include "../../profiling/all/kp_core.hpp"
 #include "kp_config.hpp"
 
+// using Kokkos::Tools::Experimental;
+// using mytpi_type = Kokkos::Tools::Experimental::ToolProgrammingInterface;
```
Comment (masterleinad) [inline common/kokkos-sampler/kp_sampler_skip.cpp]: Remove all commented lines you are not using.

File: common/kokkos-sampler/kp_sampler_skip.cpp [hunk 2]
```diff
@@ -27,6 +31,45 @@ static endFunction endForCallee                = NULL;
 static endFunction endScanCallee               = NULL;
 static endFunction endReduceCallee             = NULL;
 
+void getGlobFenceChoice() {
```
Comment (masterleinad) [inline common/kokkos-sampler/kp_sampler_skip.cpp]: The style doesn't match the rest of the file.
```suggestion
void get_global_fence_choice() {
```

File: common/kokkos-sampler/kp_sampler_skip.cpp [hunk 3]
```diff
@@ -82,9 +125,7 @@ void kokkosp_init_library(const int loadSeq, const uint64_t interfaceVer,
       printf("KokkosP: Next library to call: %s\n", nextLibrary);
       printf("KokkosP: Loading child library ..\n");
     }
-
```
Comment (masterleinad) [inline common/kokkos-sampler/kp_sampler_skip.cpp:85]: What's up with all these whitespace changes?

**Review by masterleinad — CHANGES_REQUESTED** [review]

**Comment by masterleinad** [comment]
> Sample result on MacBook using g++ (default) and Serial host backend for stream, with and without the fence.

Can you please summarize and interpret the results?

File: common/kokkos-sampler/kp_sampler_skip.cpp (hunk 1 above)
Comment (vlkale) [reply common/kokkos-sampler/kp_sampler_skip.cpp]: Thanks for that

**Comment by vlkale** [comment]
> > Sample result on MacBook using g++ (default) and Serial host backend for stream, with and without the fence.
> 
> Can you please summarize and interpret the results?

@masterleinad The result shows output of the test of the proposed new sampler with optimization to fencing, in the case that global fencing is on   (KOKKOS_TOOLS_GLOBALFENCES=1) and the case that global fences is off (KOKKOS_TOOLS_GLOBALFENCES=0). 

The output shows that given the sampling skip rate is 105, only 46 kernel invocations out of the 1000 have a Kokkos Tool child event for the Kokkos Tools connector specified (here simple-kernel-timer)  dispatched. One can intrepret this as the sampler behaving correctly, in both cases. 

Perhaps a better test case here is the output from the version of the sampler in Kokkos Tools `develop` branch and the version of sampler from this PR (again testing for both cases of globalFences turned on and global fences turned off), showing a timing comparison between the version in `develop` and this PR.  I will put this in. Also, the priority configuration of the two is the one with KOKKOS_TOOLS_GLOBALFENCES=1, i.e., the default.

File: common/kokkos-sampler/kp_sampler_skip.cpp (hunk 3 above)
Comment (vlkale) [reply common/kokkos-sampler/kp_sampler_skip.cpp:85]: I will fix, making sure the code file is indeed getting processed through clang-format.

**Comment by vlkale** [comment]
For reference, here is an output of a build and compilation on a MacBook Pro (2.9 GHz Quad-Core Intel Core i7)

```
s1017105ca:vbuild-fenceLimitedSingleLib vlkale$ cmake ..
-- 
-- ConfiguringKokkos-Tools
-- 
-- Found Kokkos installation: /usr/local
		Devices: SERIAL
		Architecture: 
		TPLs: LIBDL
		Compiler: /Library/Developer/CommandLineTools/usr/bin/c++ (AppleClang)
		CMAKE_CXX_FLAGS: 
		Options: DEPRECATED_CODE_4;DEPRECATION_WARNINGS;LAUNCH_COMPILER;IMPL_DESUL_ATOMICS;COMPLEX_ALIGN
-- PAPI support disabled
-- MPI not available. MPI disabled.
CMake Warning at cmake/configure_variorum.cmake:23 (message):
  Variorum not found: set Variorum_ROOT CMake variable or VARIORUM_ROOT
  environment variable to build Variorum connector
Call Stack (most recent call first):
  CMakeLists.txt:95 (include)


CMake Warning at CMakeLists.txt:107 (message):
  Set VTUNE_HOME in environment or VTune_ROOT in build options to build VTune
  connectors


-- Apple OSX target detected.
-- Skipping memory-hwm-mpi (MPI disabled)
-- Building Monolithic KokkosTools library with profilers: kp_kernel_timer_json;kp_kernel_timer;kp_hwm;kp_memory_events;kp_memory_usage;kp_chrome_tracing;kp_space_time_stack;kp_perfetto_connector
-- Configuring done (0.1s)
-- Generating done (0.2s)
-- Build files have been written to: /Users/vlkale/Desktop/vlap/wk/code/softwareTech/kokkos-tools-develop/vbuild-fenceLimitedSingleLib
s1017105ca:vbuild-fenceLimitedSingleLib vlkale$ make -j
[  6%] Built target kp_kokkos_sampler
[ 12%] Built target kp_kernel_shared
[ 19%] Built target kp_kernel_filter
[ 25%] Built target kp_space_time_stack
[ 32%] Built target kp_kernel_logger
[ 38%] Built target kp_memory_usage
[ 45%] Built target kp_memory_events
[ 51%] Built target kp_hwm
[ 58%] Built target kp_chrome_tracing
[ 67%] Built target kp_perfetto_connector
[ 74%] Built target kp_kernel_timer_json
[ 80%] Built target kp_kernel_timer
[ 87%] Built target kokkostools
[ 93%] Built target kp_reader
[100%] Built target kp_json_writer
s1017105ca:vbuild-fenceLimitedSingleLib vlkale$ make clean;
s1017105ca:vbuild-fenceLimitedSingleLib vlkale$ make -j
[  3%] Building CXX object common/kokkos-sampler/CMakeFiles/kp_kokkos_sampler.dir/kp_sampler_skip.cpp.o
[  6%] Building CXX object profiling/simple-kernel-timer/CMakeFiles/kp_kernel_shared.dir/kp_shared.cpp.o
[  9%] Building CXX object debugging/kernel-logger/CMakeFiles/kp_kernel_logger.dir/kp_kernel_logger.cpp.o
[ 12%] Building CXX object profiling/memory-hwm/CMakeFiles/kp_hwm.dir/kp_hwm.cpp.o
[ 16%] Building CXX object common/kernel-filter/CMakeFiles/kp_kernel_filter.dir/kp_kernel_filter.cpp.o
[ 19%] Building CXX object profiling/memory-usage/CMakeFiles/kp_memory_usage.dir/kp_memory_usage.cpp.o
[ 22%] Building CXX object profiling/memory-events/CMakeFiles/kp_memory_events.dir/kp_memory_events.cpp.o
[ 25%] Building CXX object profiling/space-time-stack/CMakeFiles/kp_space_time_stack.dir/kp_space_time_stack.cpp.o
[ 29%] Building CXX object profiling/perfetto-connector/CMakeFiles/kp_perfetto_connector.dir/libperfetto-connector.cpp.o
[ 32%] Building CXX object profiling/chrome-tracing/CMakeFiles/kp_chrome_tracing.dir/kp_chrome_tracing.cpp.o
[ 35%] Building CXX object profiling/perfetto-connector/CMakeFiles/kp_perfetto_connector.dir/perfetto/perfetto.cc.o
[ 38%] Linking CXX shared library libkp_kokkos_sampler.dylib
[ 38%] Built target kp_kokkos_sampler
[ 41%] Linking CXX shared library libkp_hwm.dylib
[ 45%] Linking CXX shared library libkp_kernel_logger.dylib
[ 45%] Built target kp_hwm
[ 48%] Linking CXX static library libkp_kernel_shared.a
[ 48%] Built target kp_kernel_logger
[ 51%] Linking CXX shared library libkp_memory_usage.dylib
[ 51%] Built target kp_kernel_shared
[ 54%] Linking CXX shared library libkp_memory_events.dylib
[ 54%] Built target kp_memory_usage
[ 58%] Building CXX object profiling/simple-kernel-timer/CMakeFiles/kp_kernel_timer_json.dir/kp_kernel_timer_json.cpp.o
[ 61%] Building CXX object profiling/simple-kernel-timer/CMakeFiles/kp_kernel_timer.dir/kp_kernel_timer.cpp.o
[ 61%] Built target kp_memory_events
[ 64%] Linking CXX shared library libkp_chrome_tracing.dylib
[ 64%] Built target kp_chrome_tracing
[ 67%] Linking CXX shared library libkp_kernel_filter.dylib
[ 70%] Linking CXX shared library libkp_kernel_timer_json.dylib
[ 70%] Built target kp_kernel_filter
[ 74%] Linking CXX shared library libkp_kernel_timer.dylib
[ 74%] Built target kp_kernel_timer_json
[ 74%] Built target kp_kernel_timer
[ 77%] Building CXX object profiling/simple-kernel-timer/CMakeFiles/kp_reader.dir/kp_reader.cpp.o
[ 80%] Building CXX object profiling/simple-kernel-timer/CMakeFiles/kp_json_writer.dir/kp_json_writer.cpp.o
[ 83%] Linking CXX shared library libkp_space_time_stack.dylib
[ 83%] Built target kp_space_time_stack
[ 87%] Linking CXX executable kp_reader
[ 87%] Built target kp_reader
[ 90%] Linking CXX executable kp_json_writer
[ 90%] Built target kp_json_writer
[ 93%] Linking CXX shared library libkp_perfetto_connector.dylib
[ 93%] Built target kp_perfetto_connector
[ 96%] Building CXX object profiling/all/CMakeFiles/kokkostools.dir/kp_all.cpp.o
[100%] Linking CXX shared library libkokkostools.dylib
[100%] Built target kokkostools
s1017105ca:vbuild-fenceLimitedSingleLib vlkale$ make install
[  6%] Built target kp_kernel_filter
[ 12%] Built target kp_kokkos_sampler
[ 19%] Built target kp_kernel_logger
[ 25%] Built target kp_kernel_shared
[ 32%] Built target kp_kernel_timer_json
[ 38%] Built target kp_kernel_timer
[ 45%] Built target kp_reader
[ 51%] Built target kp_json_writer
[ 58%] Built target kp_hwm
[ 64%] Built target kp_memory_events
[ 70%] Built target kp_memory_usage
[ 77%] Built target kp_chrome_tracing
[ 83%] Built target kp_space_time_stack
[ 93%] Built target kp_perfetto_connector
[100%] Built target kokkostools
Install the project...
-- Install configuration: ""
-- Up-to-date: /usr/local/include/kp_all.hpp
-- Up-to-date: /usr/local/include/kp_config.hpp
-- Installing: /usr/local/lib/libkokkostools.dylib
-- Installing: /usr/local/lib/libkp_kernel_shared.a
-- Installing: /usr/local/lib/libkp_kernel_timer_json.dylib
-- Installing: /usr/local/lib/libkp_kernel_timer.dylib
-- Installing: /usr/local/lib/libkp_hwm.dylib
-- Installing: /usr/local/lib/libkp_memory_events.dylib
-- Installing: /usr/local/lib/libkp_memory_usage.dylib
-- Installing: /usr/local/lib/libkp_chrome_tracing.dylib
-- Installing: /usr/local/lib/libkp_space_time_stack.dylib
-- Installing: /usr/local/lib/libkp_perfetto_connector.dylib
-- Installing: /usr/local/lib/cmake/KokkosToolsConfig.cmake
-- Installing: /usr/local/lib/cmake/KokkosToolsConfig-noconfig.cmake

```

File: common/kokkos-sampler/kp_sampler_skip.cpp (hunk 3 above)
Comment (vlkale) [reply common/kokkos-sampler/kp_sampler_skip.cpp:85]: @masterleinad  Does the new file committed address the issues you have with whitespace changes?

File: common/kokkos-sampler/kp_sampler_skip.cpp (hunk 1 above)
Comment (vlkale) [reply common/kokkos-sampler/kp_sampler_skip.cpp]: Fixed

**Comment by vlkale** [comment]
I am aware of the build test with Kokkos that's in the CI that is failing. 
Looking further into this, I found that upon testing on another machine, the Device ID is 100663297 . It should be 0. 

I don't know if that is the reason for the problem. 

I am putting this back into a draft PR to figure things out from my end.

**Comment by masterleinad** [comment]
> Looking further into this, I found that upon testing on another machine, the Device ID is 100663297 . It should be 0.

Note that the device identifier used in the `Kokkos::Tools` interface is computed as https://github.com/kokkos/kokkos/blob/4d1c6c351490cc8660e2392ead9fce1af7e379f5/core/src/impl/Kokkos_Profiling_Interface.hpp#L97-L103 and you might want ti use https://github.com/kokkos/kokkos/blob/4d1c6c351490cc8660e2392ead9fce1af7e379f5/core/src/impl/Kokkos_Profiling_Interface.hpp#L79-L86 to translate that to a device type, device id and instance id.

**Comment by vlkale** [comment]
@masterleinad Thanks for that. I will take a look at that interface.

**Comment by vlkale** [comment]
@masterleinad Thanks for that. I will take a look at that part of the interface.

**Pull request closed without merging**